File: selection/python/library/abstract_manager.py
from typing import Generic, TypeVar

from library.abstract_manager_configurations import (
    AbstractManagerConfigurations,
)
from library.error_manager import ErrorManager
from library.singleton_manager import SingletonManager


Configurations = TypeVar(
    "Configurations",
    bound=AbstractManagerConfigurations,
)

class AbstractManager(
    Generic[Configurations],
):
    def __init__(
        self,
        new_configurations: Configurations,
    ) -> None:
        self.configurations = new_configurations

    def setup(
        self,
        new_configurations: Configurations,
    ) -> bool:
        try:
            if self.configurations.is_disabled_for_handling(
                new_configurations is None,
            ):
                return False

            self.configurations = new_configurations

            return True

        except Exception as exception:
            self.handle_error_outputs(exception)

            return False

    def reset(
        self,
    ) -> bool:
        try:
            if self.configurations.is_disabled_for_handling():
                return False

            self.configurations = type(
                self.configurations,
            )()

            return True

        except Exception as exception:
            self.handle_error_outputs(exception)

            return False

    def handle_error_outputs(
        self,
        error: str | Exception,
        message: str | None = None,
    ) -> bool:
        error_manager = SingletonManager.get_singleton(
            ErrorManager,
        )

        if isinstance(error, Exception):
            return error_manager.handle_error_outputs(
                error,
            )

        if message is not None:
            return error_manager.handle_error_outputs(
                error,
                message,
            )

        return error_manager.handle_error_outputs(
            error,
        )

File: rust/src/v1/abstract_manager.rs
use std::fmt::Display;
use std::ops::{Deref, DerefMut};

use crate::v1::abstract_manager_configurations::AbstractManagerConfigurations;
use crate::v1::error_manager::ErrorManager;
use crate::v1::singleton_manager::SingletonManager;

#[derive(Debug, Clone)]
pub struct AbstractManager<T>
where
    T: Deref<Target = AbstractManagerConfigurations> + DerefMut + Default + Clone,
{
    pub configurations: T,
}

impl<T> AbstractManager<T>
where
    T: Deref<Target = AbstractManagerConfigurations> + DerefMut + Default + Clone,
{
    pub fn new(configurations: T) -> Self {
        Self { configurations }
    }

    pub fn setup(&mut self, configurations: T) -> bool {
        if self.configurations.is_disabled_for_handling(false) {
            return false;
        }

        self.configurations = configurations;

        true
    }

    pub fn reset(&mut self) -> bool {
        if self.configurations.is_disabled_for_handling(false) {
            return false;
        }

        self.configurations = T::default();

        true
    }

    pub fn handle_error<TMessage>(&self, message: TMessage) -> bool
    where
        TMessage: Display,
    {
        SingletonManager::get_singleton::<ErrorManager>()
            .handle(message)
            .is_ok()
    }

    pub fn handle_error_with_title<TTitle, TMessage>(
        &self,
        title: TTitle,
        message: TMessage,
    ) -> bool
    where
        TTitle: Display,
        TMessage: Display,
    {
        SingletonManager::get_singleton::<ErrorManager>()
            .handle_with_title(title, message)
            .is_ok()
    }
}
